### ISSUE TITLE
Implement review notes blocks doc

### ISSUE THREAD

**Issue by steersbob** [BrewBlox/brewblox-documentation#265]
File: all_blocks.md
Commit: c485bf2f804fe498ce010349ba4b5a75eb529e75
Tested on laptop with kubuntu 20.04 LTS between March 31 and April 9 2021 with Brewblox UI build date: 2021-03-30T12:16:32.267Z viewed in FireFox. Spark with Fermentation Fridge wizard blocks (standard). Spark is powered & connected over USB.

l3: Guide introduction is missing, a note is no introduction in my opinion. Most information in the note works better as an introduction:
"Blocks are combined to build control chains to run on the Brewblox Spark. In the UI, they can be displayed on a dashboard. *In this guide we will discuss all the blocks available in the Brewblox system.* For more information on blocks vs widgets, see the Blocks in depth guide." Maybe only story about reference documentation in note. After reread: example of ambiguous use of naming blocks and widgets, more later.

l14: Ah, that is a weird block with only one line of description. Oh, it is an introduction... Maybe it is more clear if the sensor types we do have are mentioned in the introduction, or the temperature sensor block titles are stripped of their heading / have a reduced heading (H3 to H4?).

Testing remark: **Blocks are tested / manipulated on the spark service page (list view)**

OneWire Temp Sensor block:
- Deleting an offset results in the offset not changed. Offset set to 0C is expected. - error notification visibility is too short to understand the error, but good that they are there 
- Why would a user want to manually change the OneWire Address of the sensor?
  - Address modification fails if the manual address is too short / long - good!
- Why would a user want to swap two OneWire Addresses? Bob: ah, to replace a broken sensor.

I removed my OneWire Temp sensor widget + block. My fermentation fridge setup is broken. Block is still there in diagram view (not in list view), but is called New|TempSensorInterface? Clicking the block makes the full UI unresponsive (see screen RemovedTempSensorDiagramViewBroken.png). Hitting F5 makes the UI responsive again, but does not fix my control chain. Let's see how I fix this... Heuj! I rediscovered my sensor, and referenced it as the sensor block from the setpoint block. All seems working again. Would be nice if you can add the sensor too by clicking the block where it is supposed to be.

Intermezzo: showing temp sensor values in a separate graph has two icons in the top-right of the graph that I have not seen before:
1 Set Y-axis range - default: auto
2 Set X-axis range through presets
Why do these show up here? They are not in the widget on the dashboard, only when the widget is shown in dialog. The options are not really needed, but since they are there it is surprising they are not shown on the dashboard.

l26: fluctuations option of the Mock Temp Sensor not mentioned in the guide.

l30: maybe discuss some use cases of combined sensors? I can imagine none... Have multiple sensors measure your fridge air temperature and control the average? Better mix the air with a fan...

Temp Sensor (Combined) block:
- it is confusing the view options in the top-right are still called widgets. "Show full widget" / "Show basic widget" where "Show full block" / "Show basic block" is expected. In list view show full block / show basic block actually makes sense. For widgets on the dashboard not so much.
- Minor: In the change block name menu, the OK button is unusable without changing the block name. You don't want to change the block name into an existing name, but it would be nice if the OK button is usable if the name is not changed.
- Minor: the show on dashboard menu has no default selection, even if there is only one dashboard. You have to select the dashboard to be able to hit OK.
- Funny: "Show in dialog" shows the full view in the list view, while clicking a block in the diagram view shows the basic view of the same dialog. Makes sense for a Setpoint or PID block, not so much for a sensor block

l33: Setpoint, sensor and PID are all blocks but not named that way. For consistency and clarity we better make a choice, the same with capitalization of block names; Setpoint and PID are capitalized, sensor is not.

l38: "The filter can be bypassed to respond *immediately* when...sets the difference at which *the filter is bypassed*."

Setpoint block:
- List view: basic / full block view status not remembered upon forced reload & menu switch (Services <-> Dashboard)
- Modify filter period menu:
   - spelling error: "Changes faster than the filter period will be filtERed out."
   - No arbitrary filter periods available? Is there a technical reason for this? Due to the historyDB?

l44: "You configure *the profile*..."

l48: ".., the Setpoint Profile is working its way from point 1 to 2,..."

l51: "..so you can easily re-use profiles *for future brew sessions*."

l58: On hold until the Spark reconnects to what? To the server? What will the Spark do once it knows the date & time again? Continue where it stopped? Continue from the actual time if listed in the profile?

Setpoint Profile block:
- Start time menu: 
  - description of date format is capitalized YYYY/MM/DD, while time format is a mix HH:mm:ss. Offset notation is all     lower case; smhd. Everything lower case is best probably.
  - menu text: "The offset for the first point is always 0s." This is not enforced, I can manually shift the offset of the first point by changing the offset itself, or setting the absolute time later than the start time. By forcing a profile shift (changing the start time) the time offset of the first point is *not* reset to 0s either. It seems like the first point of the series is the start time and the temperature value of the start time point is whatever is set at the next point.
- I like the graph + menu side by side in the dialog to see what I do
- Graph:
  - Now indicator (vertical dashed green line) is not described in the legend or anywhere else.
  - Spike lines don't work, whats the use of the select tools? Compare data on hover not useful
  - What's the difference between autoscale and reset axes?

General remark: selecting one or multiple blocks in list view creates an unnecessary vertical scroll bar in the blocks column on the right side of the window - see ListViewUnnecessaryScrollbar.png

l60: it takes a while to understand what a digital actuator is, and what the digital actuator block is for. Consider rewritten piece:

"Actuators are hardware elements that act on things in the real world, like temperature or water flow. Digital actuators have two states; ON and OFF. Examples are: a heater, cooler, pump or a valve. To control digital actuators, a *Digital Actuator* block is used in the Brewblox control chain.

A *Digital Actuator* block is a software block that manages the state of a single digital output pin. Digital output pins are hardware elements that are detected as a separate block. In the digital actuator block you specify the output pin and set constraints if required. The available constraint types are: *Minimum ON time*, *Minimum OFF time*, and *Mutually exclusive*. 

A few blocks provide digital output pins.
These blocks are *Spark Pins*, *DS2413 Chip*, and *DS2408 Chip*. 
They are automatically detected and represent hardware elements."

Remarks about the piece above:
- weird that the hardware blocks are discussed in the piece after l60 under separate headings. The digital actuator block itself has no separate heading
- constraints are mentioned, but discussed in a separate chapter. You have to search for that or find out when you read the full piece.
- Maybe best to group all the actuators in one section. Make the distinction between digital and analog actuators in the introduction with examples. Then discuss digital actuators + constraints, followed by analog actuators + constraints.

Digital Actuator block:
- Edit constraints menu: 
  - title Extra lock time setting not fully visible; visible is only: Extra loc... due to "Use default value from Mutex block" icon. It seems the block size is dependent on the length of the Mutex name. "Ferment Mutex" in example from Fridge Wizard.
  - "Use default value from Mutex block" icon does not show up when you manually add a mutex constraint. - Oh, it does when reopening the edit constraints menu
  - Adding a mutex constraint without defining the mutex block does not trigger a warning. It does prevent the digital actuator block switching to the ON state 

l80: You can only see the current voltages of the power supply when you open "Show full widget"

SparkPins block:
- weird, Spark is connected over USB. 5 & 12V are enabled. 5V reading is 9.68V, 12V reads 4.11V see SparkVoltageReading.png. Swapped readings? Similar behavior when the Spark is connected to a separate USB adapter & connected over WiFi.
- You are not prevented from renaming the block. Once renamed, reverting to the default name is prohibited as SparkPins is a protected name.

Diagram view:
- Situation: Digital actuator positioned at the top with two connections below: SparkPins and Mutex. Removing the Mutex leaves a block with a non-existent standard name. Removing the Digital Actuator forces diagram rebuild, and removes the non-existent block.

DS2413 Chip:
- works & block works

DS2408 Chip:
- cannot test, board absent

Motor Valve block:
- works as far as I can test (DS2408 Chip absent), gives warning about no selected channel
- does not work with mock pins block
- can be driven with logic actuator discussed later

l107: "The *Mock Pins* block has 8 virtual pins that can be set as a target for a Digital Actuator. The block can be used for testing, or serves as an output for custom logic built with the Logic Actuator block discussed later."

l111: you talk about a PID before the block is introduced.

l113: "..actuatorS:"

l129: confusing formatting of example. The example is best formatted in plain text with a bold **Example:** heading. The note of l137 is better formatted as a tip (:::tip, etc.). I feel the PWM concept can be explained more clearly, but lack the skill and experience to do so. Isn't it better to call the period the *PWM period*?

PID block:
- tested with Cool PID of ferment fridge wizard
- in "Show full widget" mode in list view the 'add bar' (____+_) breaks the PID block boundary due to small screen width (1280x1024), see PIDlistView_full.png. Behavior is gone on bigger screen (1440x900) or when side panel is collapsed.
- In list view "Show full widget" mode reset to basic after diagram view is toggled, other menu is toggled, or hard refresh (F5) is forced
- cannot really test functionality due to lack of setup, all buttons work and can be manipulated without error
- Minor detail: in list view / full widget mode, PWM inactive warning shortly visible after manually enabling PID

l141: "Sometimes a temperature is best controlled indirectly by managing another temperature. The *Setpoint Driver* block turns the difference between two temperatures into an actuator.

An example of the use of a setpoint driver can be found in a HERMS system. In a HERMS system, the mash is heated indirectly by a heating element in a separate hot water tun. Lets look at the example."

l148: is it possible to add labels to the figure? HLT, MT, how is the right tun called? Boil Tun (BT)?

l159: Example image of the 2 PID control chains maybe?

l169: Remarks:
1.  Graph line color legend encoded in text
2.  what determines the maximum HLT temperature? Denaturation of the mash enzymes I presume? Can you heat to boil? I think not but the mash will only be that hot for a short period... When does it start to influence the mash sugar composition? Questions...

HERMS example:
- cannot test HERMS example due to lack of sensors. 3 temperature sensors are required, I have 2 available.
- Curious: how do you trigger the setup wizard when you already have a dashboard? Workaround: I created a new dashboard with the +-button of the dashboard list in the side panel. The empty dashboard had the get started button I needed. Ah! There is a 'Start a wizard' button in the admin panel.
- Pointing the HERMS wizard to an already created dashboard: I created a new dashboard with the +-button in the side panel. Named it herms, and started the wizard with the get started button. In the wizard the Unique dashboard ID was set to herms-2. The herms ID was already taken by the newly created, empty dashboard. Cannot set the previously created dashboard as the target. Should have skipped the new dashboard and used the route via the admin panel.

Setpoint Driver block:
- I can test the Setpoint Driver block by manually rebuilding the 'fermentation fridge with a dynamic setpoint driver' example from the control chains guide
- Services - list view: setpoint driver block gets deselected after renaming and is removed from the column on the right. Same behavior for other blocks
- Curious - not really important: Services - diagram view: view gets chaotic after building dynamic fridge setpoint into control chain of fermentation fridge setup, see DiagramViewBuildingDynamicFridgeSetpoint.png. The logic actuator link to the Ferment cool actuator is the culprit. Removing the link sanitizes the diagram view, see DiagramViewBuildingDynamicFridgeSetpoint_LAlinkRemoved.png. Removing the link between the LA block and the Motor Valve block works too, removing the Fridge setpoint profile works too.
- Full widget view: What does the 'Reference field' setting mean? The setting has two options; Setting / Measured. Option Setting is default. Do you have to tell the setpoint driver when it is linked to a measured value (sensor)? In what circumstances would you want that, is there an example somewhere?
- all buttons / settings work as far as I can test

l180: add an example of the use of an analog actuator maybe?

Analog Actuator (Mock) block:
- full view: entering a setting outside clip to min/max values does not trigger a warning. Neither does triggering the clip to min/max value settings. Color coding the triggered clip could help too.
- Why can you add min / max value constraints when you already have the Clip to min / max value options? They do work though

l187: separate into two sentences: "How it works: a digital actuator has a *desiredState* and a *state*. The *desiredState* is set by you or the PWM block driving the block, the *state* is the actual status of the physical pin."

Digital Actuator block - constraints test:
- ON / OFF status update dependent on list view update. When waiting on the minimum on / off time constraint (spinner) you actually wait on the constraint time + list view update time. You can actually change the actuator state within the constraint time, neat!
- constraints work

General remark: Mutex gets a separate section. I agree Minimum ON time, Minimum OFF time, Delay ON and Delay OFF are trivial, but it would be nice to officially define them or at least name that their function follows the naming

l213: "..are driven by *a PWM block*, the..."

l216: "..add a Balancer *block* (see below)."

Services - List view:
- Curious behavior: selecting a block for view in the right column moves the just selected block to the top of the view. Previously selected blocks can move out of view, even if there are no blocks below the block just selected. Newly added blocks are added to the view list too, and can result in the same behavior.

l221: although partially obvious I would like to know what the constraints entail. The balancer is the only constraint discussed in detail.

l223: confusing to talk about values while one of the values is called *value*. "Analog actuator constraints operate upon three actuator properties: *desiredSetting*..."

l230: constraints of a PWM block target, what are those? Like a Mutex?

l231: "..uses the measured (filtered?) value of the setpoint *as a reference*,..."

l236: add "To prevent one element blocking the second to turn on for a considerable amount of time."

l240: oh, example explained later

Balancer block - tested on 2 mock Analog Actuators:
- Analog actuators have no target block option to make a fully functional mock test of the Mutex - Balancer interaction
- Adding a Balancer constraint to two mock Analog Actuators renders the setting / value readout of the actuators useless (block full view). It is set to 0, or some clipped value, see AnalogActuatorBalancer.png. What is the desired behavior?
- Curious: if two analog actuators ask for a sum of less than 100%, the balancer grants more than requested. Request 0 and 90, Granted 5 and 95. Is this desired behavior? Don't you at least want the option to grant less than 100% when that is the total sum of the requests?

Blancer block - tested on fermentation fridge setup:
- all seems to work as far as I can test

l260: you introduce the settings by stating there are three sections, I can imagine those from the picture but the headings do not follow the sections. I was expecting sections: Input / Output, PID, Boil Mode. 

l264: you can also directly edit the setting of the driven analog actuator block

l266: PID section starts here, but is not marked by heading. Propose H4 heading for PID, then Proportional, Integral and Derivative in **bold**.

l272: "agGressive"

l299: how does the PID know the actuator cannot achieve the output value? Do we assume every PID to control the temperature of a watery substance? How is this called anti-windup? The Wiki article on PID Integral windup (https://en.wikipedia.org/wiki/Integral_windup) deals with integral overshoot due to a sudden setpoint change...

l311: which filtered value of the input? Is this the filtered sensor value from the setpoint?

Remarks:
- Ok, now you have some PID calculated output value. How does this get interpreted by a PWM to a value in the 0 - 100% range? Theoretically the output of a PID calculation has no limit... The PWM block section does not say anything about this either...
- Do we have a guide about PID tuning? If so, might be good to put a link here...

l313: Boil mode options *Minimum output when boiling* and *Boil temperature setting* are only introduced implicitly in the story.

PID block:
- tested with fermentation fridge setup from wizard
- I was expecting the boil mode setting order to be reversed. First set the boil temperature (left), then the minimum output (right).
- why is the Boil temperature setting implemented as an offset from 100C? Is this for 'Mericans who cannot remember the Fahrenheit boiling temperature of water? Maybe a better name is: *Boil temperature offset*?
  - Negative temperature offset of -2C leads to ugly setting view: 100+**-2.00**C. Maybe evaluate the value (98C) or lose the static +-sign and print the sign of the offset in all cases
  - Idea: can the Boil temperature setting availability be made dependent on the sign of Kp? Who needs a boil temperature setting when the PID controls a cooling actuator... Or do we envisage situations where you want water to be and stay frozen solid? *Freeze mode*
  - I thought we wanted Brewblox to be modular and generic for use outside of breweries. We do slip in all this brewery functions and naming which makes it less generic.

l331: suggestion: "..the Logic Actuator *is meant to take care of* additional behavior.."

l334: Multiple fermenters sharing a glycol pump. Each individual fermenter *has* a valve connecting its own coil *to* a shared loop. The pump should be active *when any valve* is open.

Consistency: l335 - 337 Turning *ON* (3x)

l347: Maybe add a line about Digital Actuators getting lower case labels, Analog Actuators getting upper case. - Or maybe add it to l366.

l357: Not really clear this is a list of examples. Box maybe? **If you want to check ... create a comparison for a ... *block***

l368: The NOT operator is the only one with an explicit explanation in 'true / false form' !true = false, !false = true. Listed examples are only example expressions, not their meaning. Maybe add explanation to other operators too. true & true = true, true & false = false, true | true = true, true | false = true, false | false = false, true ^ true = false, true ^ false = true, false ^ false = false, brackets example: true & (true | false) = true - weird example...

l383: Ah, more examples here. Explicit examples of OR and AND operations and brackets are described, XOR is missing. Maybe box the example to stand out from the other text.

l402:
- Content: This is an example, but not named or formatted that way.
- Text - better describe the situation: "In a fermentation fridge setup with both a heater and a cooler, you might want to add a fan to mix the fridge air and even out temperature gradients. You can use the Logic Actuator to drive the fan. Add an expression like `a|b`, to check that the heater or cooler is ON. The logic actuator will set the desired state of the fan Digital Actuator. Next you can set a `Delayed ON` and a `Delayed OFF` constraint on that fan actuator. This would cause the fan to start a while after the heater/cooler has turned ON and run for a while after it has turned OFF."

Logic Actuator block:
- I manually rebuilt the setup shown in the guide for testing; 1 Logic Acutuator, 4 Digital Actuators, 1 setpoint and 1 PWM.
- Minor: full view - starting from a new block the Expression field shows an example (a|b)&(c^d)&!e. There is no interaction possible with the example. It will disappear when typing in the field
- Boolean logic checks when typing the example are neat!
- Red / Green Status dots of Active comparisons do not change color - presumably repaired in new release. Comparison labels (a, b, A, B) *do* change color in the expression field
- Removing a comparison also removes its label (B) from the expression, even if a later added comparison (C) now has the label (B) - neat! Similar: labels in the expression change when a comparison gets another label.
- Minor: there is no option to reorder the comparisons
- What is the logic behind lower / upper case labeling of comparisons? Digital Actuators get lower case labels, Analog Actuators get upper case?
- Setpoint driver block not available to logic actuator. I cannot think of an useful example, but might be there.
- operators for PWM comparisons (Edit comparison menu) are called *Measured value* and *Setting* in Brewblox UI and *measured state* / *desired state* in the guide. I think the latter is correct
- It is not clear from the menu that the target value for a PWM is a percentage while it is clear that the target value of a setpoint is a temperature by the Celsius symbol. The percentage symbol is added in the comparison list.
- operators for comparisons are called *measured value* and *measured state* where they are called *value* and *state* in the description actuator constraint sections of the blocks. Is there a reason for this inconsistency? Might lead to confusion.
- Edit comparison menu - UI inconsistency: The 'target value setting' in edit comparison menu of Digital Actuators is not explicitly named in the UI, only a graphic ON/OFF toggle. This way the use of the toggle is not clear explicitly. The Operator / Target value labels of the Analog Actuators are more clear.

Display Settings block:
- sensors and setpoints get an icon, PIDs and PWMs get a status dot. Satus dot is needed, but are there no icons?
- you have to figure out the meaning of the data on the screen yourself
  - sensor: one value = temperature, clear - add C/F indication? - icon is neat
  - setpoint: two values, unclear which one is which, Funny the block is called setpoint and the measured value is represented bigger. Icon is neat - maybe add an arrow next to the setpoint value?
  - PID: too bad no Icon, would not know one either. I assume the bars represent P, I and D values currently evaluated? Printed values are setpoint, desired percentage, measured value and achieved percentage? - I have to guess
  - PWM: too bad no icon. Idea for an icon: pulse train?. Shown values are desired and achieved percentage I presume. Dunno which one is which.
  - What is all the information in the top bar? Connected to USB, Connected to WiFi & IP, percentage of processor use, percentage memory use? I have to guess for the latter two.
- General remarks:
  - would be nice if there is an explanation somewhere of what everything on the Spark display means, and where all the blocks are explained. 
  - what is the use of the PWM block? All information is already listed in the PID block. I cannot imagine a case where you have a PWM block without a PID block, correct? The same for the setpoint block. What is a setpoint without a PID block? Cool that the blocks are there, maybe not needed.
- Minor - Block color menu: entry type (gamut, sliders, presets) is not remembered upon menu exit. Annoying for recurrent visits. Color is only set after hitting OK. Maybe add a 'set color button' (SET) to prevent recurrent visits to the menu.

Quirk: all localhosts follow the same zoom setting. When viewed in Firefox in two separate windows, the Brewblox UI running on localhost, and the Brewblox documentation running on localhost:8080 both zoom in/out when changing the zoom setting of either one. Ah, presumably it's a Firefox thing. This happens to multiple tabs of the same domain too.

Updating to Brewblox release 2021/04/07, Build: 2021-04-07T14:53:23.863Z, for testing of updates on Logic Actuator and DisplaySettings changes.

Notes:
- (feature) Added global configuration setting for preferred temperature unit (Celsius / Fahrenheit). This setting is used by the UI and all Spark and Tilt services.
  - works!
  - Notification is persistent
  - Wonder whether the notification will return after future updates. It is described like that in the release notes introduction...
- (fix) The dot indicators in the Logic Actuator widget now correctly toggle between red/green to show condition result.
  - works!
